packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -110,6 +111,41 @@ function prepareTree(root: IConstruct) {
   visit(root, 'post', construct => construct.onPrepare());
 }
 
+/**
+ * Find all stacks and add Metadata Resources to all of them
+ *
+ * There is no good generic place to do this. Can't do it in the constructor
+ * (because adding a child construct makes it impossible to set context on the
+ * node), and the generic prepare phase is deprecated.
+ *
+ * Only do this on [parent] stacks (not nested stacks), don't do this when
+ * disabled by the user.
+ *
+ * Also, only when running via the CLI. If we do it unconditionally,
+ * all unit tests everywhere are going to break massively. I've spent a day
+ * fixing our own, but downstream users would be affected just as badly.
+ *
+ * Stop at Assembly boundaries.
+ */
+function injectMetadataResources(root: IConstruct) {
+  visit(root, 'post', construct => {
+    // Only on top-level stacks
+    if (!Stack.isStack(construct)
+      || construct.parentStack
+      // Unless disabled
+      || construct.node.tryGetContext(cxapi.DISABLE_VERSION_REPORTING)
```

Comment:
Should this be a property of Stack such as `stack.versionReportingEnabled` which will take into account context and the nested stack posture instead of leaking the fact that this _can_ he set through context?

Reply:
I was also thinking that this should be an opt-in switch ("enable") and not opt-out ("disable"). This way, we (and our users) won't need to fix all their unit tests because the metadata resource will only be attached if the CLI sets that context flag.

Reply:
We could do that, but since we already have an existing opt-out flag: `DISABLE_VERSION_REPORTING` (which we need to support for backwards compatibility anyway) I didn't really want to add another `ENABLE_VERSION_REPORTING` flag as well.

The `Stack` property probably makes sense although it would need to be exposed as a `public` property or a (ew!) `public @internal` one so that this function external to the class can use it.

It fights with the App's `{ runtimeInfo?: @default true }` property as well...

Reply:
`DISABLE_VERSION_REPORTING` is not really public API, so I would recommend removing it in favor of `ENABLE_VERSION_REPORTING` and deprecate `runtimeInfo` in favor of `versionReporting` (in props and at the `Stack` level as public API.